File: lib/data/local/app_database.dart
import 'dart:io';

import 'package:drift/drift.dart';
import 'package:drift/native.dart';
import 'package:sqlite3/sqlite3.dart' as sqlite;
import 'package:uuid/uuid.dart';

part 'app_database.g.dart';

@DataClassName('TodoItem')
class Todos extends Table {
  TextColumn get id => text()();
  TextColumn get date => text()();
  TextColumn get title => text().withLength(min: 1, max: 80)();
  TextColumn get note =>
      text().withLength(max: 300).withDefault(const Constant(''))();
  BoolColumn get isCompleted => boolean().withDefault(const Constant(false))();
  TextColumn get sortOrder => text()();
  IntColumn get createdAt => integer()();
  IntColumn get updatedAt => integer()();
  IntColumn get deletedAt => integer().nullable()();
  TextColumn get deviceId => text()();
  IntColumn get revision => integer().withDefault(const Constant(1))();

  @override
  Set<Column<Object>> get primaryKey => {id};
}

class DailyReviews extends Table {
  TextColumn get id => text()();
  TextColumn get date => text().unique()();
  TextColumn get highlight =>
      text().withLength(max: 1000).withDefault(const Constant(''))();
  TextColumn get challenge =>
      text().withLength(max: 1000).withDefault(const Constant(''))();
  TextColumn get learning =>
      text().withLength(max: 1000).withDefault(const Constant(''))();
  TextColumn get tomorrowFocus =>
      text().withLength(max: 1000).withDefault(const Constant(''))();
  IntColumn get mood => integer().withDefault(const Constant(0))();
  IntColumn get createdAt => integer()();
  IntColumn get updatedAt => integer()();
  IntColumn get deletedAt => integer().nullable()();
  TextColumn get deviceId => text()();
  IntColumn get revision => integer().withDefault(const Constant(1))();

  @override
  Set<Column<Object>> get primaryKey => {id};
}

class DailySummaries extends Table {
  TextColumn get id => text()();
  TextColumn get date => text().unique()();
  TextColumn get content => text()();
  TextColumn get provider => text()();
  TextColumn get model => text()();
  IntColumn get generatedAt => integer()();
  IntColumn get updatedAt => integer()();
  IntColumn get deletedAt => integer().nullable()();
  TextColumn get deviceId => text()();
  IntColumn get revision => integer().withDefault(const Constant(1))();

  @override
  Set<Column<Object>> get primaryKey => {id};
}

class AppMetadata extends Table {
  IntColumn get id => integer().withDefault(const Constant(1))();
  TextColumn get deviceId => text()();
  IntColumn get schemaVersionValue => integer()();
  IntColumn get legacyMigrationVersion =>
      integer().withDefault(const Constant(0))();

  @override
  Set<Column<Object>> get primaryKey => {id};
}

@DriftDatabase(tables: [Todos, DailyReviews, DailySummaries, AppMetadata])
class AppDatabase extends _$AppDatabase {
  AppDatabase(super.executor);

  static const _uuid = Uuid();

  static Future<AppDatabase> open(String path) async {
    final file = File(path);
    if (await file.exists()) {
      final database = sqlite.sqlite3.open(path);
      final version = database.userVersion;
      if (version == 1 || version == 2) {
        database.execute('PRAGMA wal_checkpoint(FULL)');
        database.close();
        final backup = File('$path.room-v$version.backup');
        if (!await backup.exists()) {
          await file.copy(backup.path);
        }
      } else {
        database.close();
      }
    } else {
      await file.parent.create(recursive: true);
    }
    return AppDatabase(NativeDatabase.createInBackground(file));
  }

  @override
  int get schemaVersion => 3;

  @override
  MigrationStrategy get migration => MigrationStrategy(
    onCreate: (migrator) async {
      await migrator.createAll();
      await _createIndexes();
      await _ensureMetadata();
    },
    onUpgrade: (migrator, from, to) async {
      if (from < 3) {
        await _migrateLegacyRoom(migrator, from);
      }
    },
    beforeOpen: (details) async {
      await customStatement('PRAGMA foreign_keys = ON');
      await _createIndexes();
      await _ensureMetadata();
    },
  );

  Future<void> _createIndexes() async {
    await customStatement(
      'CREATE INDEX IF NOT EXISTS todos_date_idx ON todos(date)',
    );
    await customStatement(
      'CREATE INDEX IF NOT EXISTS reviews_date_idx ON daily_reviews(date)',
    );
    await customStatement(
      'CREATE INDEX IF NOT EXISTS summaries_date_idx '
      'ON daily_summaries(date)',
    );
  }

  Future<void> _ensureMetadata() async {
    final current = await select(appMetadata).getSingleOrNull();
    if (current == null) {
      await into(appMetadata).insert(
        AppMetadataCompanion.insert(
          deviceId: _uuid.v4(),
          schemaVersionValue: schemaVersion,
          legacyMigrationVersion: const Value(0),
        ),
      );
    }
  }

  Future<bool> _tableExists(String table) async {
    final row = await customSelect(
      "SELECT name FROM sqlite_master WHERE type='table' AND name = ? LIMIT 1",
      variables: [Variable.withString(table)],
    ).getSingleOrNull();
    return row != null;
  }

  Future<void> _migrateLegacyRoom(Migrator migrator, int from) async {
    final hasTodos = await _tableExists('todos');
    final hasReviews = await _tableExists('daily_reviews');
    final hasSummaries = await _tableExists('daily_summaries');

    if (hasTodos) {
      await customStatement('ALTER TABLE todos RENAME TO legacy_todos');
    }
    if (hasReviews) {
      await customStatement(
        'ALTER TABLE daily_reviews RENAME TO legacy_daily_reviews',
      );
    }
    if (hasSummaries) {
      await customStatement(
        'ALTER TABLE daily_summaries RENAME TO legacy_daily_summaries',
      );
    }

    await migrator.createAll();
    final deviceId = _uuid.v4();

    if (hasTodos) {
      final rows = await customSelect('SELECT * FROM legacy_todos').get();
      for (final row in rows) {
        final createdAt = row.read<int>('createdAt');
        await into(todos).insert(
          TodosCompanion.insert(
            id: _uuid.v4(),
            date: row.read<String>('date'),
            title: row.read<String>('title'),
            note: Value(row.read<String>('note')),
            isCompleted: Value(row.read<int>('isCompleted') != 0),
            sortOrder: '${createdAt.toString().padLeft(20, '0')}-legacy',
            createdAt: createdAt,
            updatedAt: createdAt,
            deviceId: deviceId,
          ),
        );
      }
      await customStatement('DROP TABLE legacy_todos');
    }

    if (hasReviews) {
      final rows = await customSelect(
        'SELECT * FROM legacy_daily_reviews',
      ).get();
      for (final row in rows) {
        final updatedAt = row.read<int>('updatedAt');
        await into(dailyReviews).insert(
          DailyReviewsCompanion.insert(
            id: _uuid.v4(),
            date: row.read<String>('date'),
            highlight: Value(row.read<String>('highlight')),
            challenge: Value(row.read<String>('challenge')),
            learning: Value(row.read<String>('learning')),
            tomorrowFocus: Value(row.read<String>('tomorrowFocus')),
            mood: Value(row.read<int>('mood')),
            createdAt: updatedAt,
            updatedAt: updatedAt,
            deviceId: deviceId,
          ),
        );
      }
      await customStatement('DROP TABLE legacy_daily_reviews');
    }

    if (hasSummaries) {
      final rows = await customSelect(
        'SELECT * FROM legacy_daily_summaries',
      ).get();
      for (final row in rows) {
        final generatedAt = row.read<int>('generatedAt');
        await into(dailySummaries).insert(
          DailySummariesCompanion.insert(
            id: _uuid.v4(),
            date: row.read<String>('date'),
            content: row.read<String>('content'),
            provider: row.read<String>('provider'),
            model: row.read<String>('model'),
            generatedAt: generatedAt,
            updatedAt: generatedAt,
            deviceId: deviceId,
          ),
        );
      }
      await customStatement('DROP TABLE legacy_daily_summaries');
    }

    await into(appMetadata).insert(
      AppMetadataCompanion.insert(
        deviceId: deviceId,
        schemaVersionValue: schemaVersion,
        legacyMigrationVersion: Value(from),
      ),
    );
    await _createIndexes();
  }
}

File: app/src/main/java/com/limeday/app/data/AppDatabase.kt
package com.limeday.app.data

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import java.util.UUID

@Database(
    entities = [TodoItem::class, DailyReview::class, DailySummary::class, AppMetadata::class],
    version = 5,
    exportSchema = true
)
abstract class AppDatabase : RoomDatabase() {
    abstract fun limeDayDao(): LimeDayDao

    companion object {
        @Volatile private var instance: AppDatabase? = null

        fun getInstance(context: Context): AppDatabase = instance ?: synchronized(this) {
            instance ?: Room.databaseBuilder(
                context.applicationContext,
                AppDatabase::class.java,
                "lime_day.db"
            ).addMigrations(MIGRATION_1_4, MIGRATION_2_4, MIGRATION_3_4, MIGRATION_4_5)
                .addCallback(object : Callback() {
                    override fun onCreate(db: SupportSQLiteDatabase) {
                        super.onCreate(db)
                        insertMetadata(db, legacyVersion = 0)
                    }
                })
                .build()
                .also { instance = it }
        }

        internal val MIGRATION_1_4 = object : Migration(1, 4) {
            override fun migrate(db: SupportSQLiteDatabase) = migrateRoomLegacy(db, 1)
        }

        internal val MIGRATION_2_4 = object : Migration(2, 4) {
            override fun migrate(db: SupportSQLiteDatabase) = migrateRoomLegacy(db, 2)
        }

        internal val MIGRATION_3_4 = object : Migration(3, 4) {
            override fun migrate(db: SupportSQLiteDatabase) = migrateDrift(db)
        }

        internal val MIGRATION_4_5 = object : Migration(4, 5) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("ALTER TABLE todos ADD COLUMN priority INTEGER NOT NULL DEFAULT 1")
                db.execSQL("UPDATE app_metadata SET schema_version_value = 5 WHERE id = 1")
            }
        }

        private fun migrateRoomLegacy(db: SupportSQLiteDatabase, version: Int) {
            db.execSQL("ALTER TABLE todos RENAME TO legacy_todos")
            db.execSQL("ALTER TABLE daily_reviews RENAME TO legacy_daily_reviews")
            if (version >= 2) db.execSQL("ALTER TABLE daily_summaries RENAME TO legacy_daily_summaries")
            createV4Schema(db)
            val deviceId = UUID.randomUUID().toString()
            db.execSQL(
                """INSERT INTO todos (id, date, title, note, is_completed, sort_order, created_at, updated_at, deleted_at, device_id, revision)
                    SELECT lower(hex(randomblob(4))) || '-' || lower(hex(randomblob(2))) || '-4' || substr(lower(hex(randomblob(2))),2) || '-' || substr('89ab',abs(random()) % 4 + 1,1) || substr(lower(hex(randomblob(2))),2) || '-' || lower(hex(randomblob(6))),
                    date, title, note, isCompleted, printf('%020d-legacy', createdAt), createdAt, createdAt, NULL, ?, 1
                    FROM legacy_todos""".trimIndent(),
                arrayOf(deviceId)
            )
            db.execSQL(
                """INSERT INTO daily_reviews (id, date, highlight, challenge, learning, tomorrow_focus, mood, created_at, updated_at, deleted_at, device_id, revision)
                    SELECT lower(hex(randomblob(4))) || '-' || lower(hex(randomblob(2))) || '-4' || substr(lower(hex(randomblob(2))),2) || '-' || substr('89ab',abs(random()) % 4 + 1,1) || substr(lower(hex(randomblob(2))),2) || '-' || lower(hex(randomblob(6))),
                    date, highlight, challenge, learning, tomorrowFocus, mood, updatedAt, updatedAt, NULL, ?, 1
                    FROM legacy_daily_reviews""".trimIndent(),
                arrayOf(deviceId)
            )
            if (version >= 2) {
                db.execSQL(
                    """INSERT INTO daily_summaries (id, date, content, provider, model, generated_at, updated_at, deleted_at, device_id, revision)
                        SELECT lower(hex(randomblob(4))) || '-' || lower(hex(randomblob(2))) || '-4' || substr(lower(hex(randomblob(2))),2) || '-' || substr('89ab',abs(random()) % 4 + 1,1) || substr(lower(hex(randomblob(2))),2) || '-' || lower(hex(randomblob(6))),
                        date, content, provider, model, generatedAt, generatedAt, NULL, ?, 1
                        FROM legacy_daily_summaries""".trimIndent(),
                    arrayOf(deviceId)
                )
            }
            db.execSQL("DROP TABLE legacy_todos")
            db.execSQL("DROP TABLE legacy_daily_reviews")
            if (version >= 2) db.execSQL("DROP TABLE legacy_daily_summaries")
            insertMetadata(db, deviceId, version, schemaVersion = 4)
        }

        private fun migrateDrift(db: SupportSQLiteDatabase) {
            db.execSQL("ALTER TABLE todos RENAME TO drift_todos")
            db.execSQL("ALTER TABLE daily_reviews RENAME TO drift_daily_reviews")
            db.execSQL("ALTER TABLE daily_summaries RENAME TO drift_daily_summaries")
            db.execSQL("ALTER TABLE app_metadata RENAME TO drift_app_metadata")
            db.execSQL("DROP INDEX IF EXISTS todos_date_idx")
            db.execSQL("DROP INDEX IF EXISTS reviews_date_idx")
            db.execSQL("DROP INDEX IF EXISTS summaries_date_idx")
            createV4Schema(db)
            db.execSQL("""INSERT INTO todos SELECT id, date, title, note, is_completed, sort_order, created_at, updated_at, deleted_at, device_id, revision FROM drift_todos""")
            db.execSQL("""INSERT INTO daily_reviews SELECT id, date, highlight, challenge, learning, tomorrow_focus, mood, created_at, updated_at, deleted_at, device_id, revision FROM drift_daily_reviews""")
            db.execSQL("""INSERT INTO daily_summaries SELECT id, date, content, provider, model, generated_at, updated_at, deleted_at, device_id, revision FROM drift_daily_summaries""")
            db.execSQL(
                """INSERT INTO app_metadata (id, device_id, schema_version_value, legacy_migration_version, last_sync_at, last_sync_status)
                    SELECT id, device_id, 4, legacy_migration_version, NULL, '' FROM drift_app_metadata""".trimIndent()
            )
            db.execSQL("DROP TABLE drift_todos")
            db.execSQL("DROP TABLE drift_daily_reviews")
            db.execSQL("DROP TABLE drift_daily_summaries")
            db.execSQL("DROP TABLE drift_app_metadata")
        }

        private fun createV4Schema(db: SupportSQLiteDatabase) {
            db.execSQL("""CREATE TABLE IF NOT EXISTS todos (id TEXT NOT NULL, date TEXT NOT NULL, title TEXT NOT NULL, note TEXT NOT NULL DEFAULT '', is_completed INTEGER NOT NULL DEFAULT 0, sort_order TEXT NOT NULL, created_at INTEGER NOT NULL, updated_at INTEGER NOT NULL, deleted_at INTEGER, device_id TEXT NOT NULL, revision INTEGER NOT NULL DEFAULT 1, PRIMARY KEY(id))""")
            db.execSQL("CREATE INDEX IF NOT EXISTS todos_date_idx ON todos(date)")
            db.execSQL("""CREATE TABLE IF NOT EXISTS daily_reviews (id TEXT NOT NULL, date TEXT NOT NULL, highlight TEXT NOT NULL DEFAULT '', challenge TEXT NOT NULL DEFAULT '', learning TEXT NOT NULL DEFAULT '', tomorrow_focus TEXT NOT NULL DEFAULT '', mood INTEGER NOT NULL DEFAULT 0, created_at INTEGER NOT NULL, updated_at INTEGER NOT NULL, deleted_at INTEGER, device_id TEXT NOT NULL, revision INTEGER NOT NULL DEFAULT 1, PRIMARY KEY(id))""")
            db.execSQL("CREATE UNIQUE INDEX IF NOT EXISTS reviews_date_idx ON daily_reviews(date)")
            db.execSQL("""CREATE TABLE IF NOT EXISTS daily_summaries (id TEXT NOT NULL, date TEXT NOT NULL, content TEXT NOT NULL, provider TEXT NOT NULL, model TEXT NOT NULL, generated_at INTEGER NOT NULL, updated_at INTEGER NOT NULL, deleted_at INTEGER, device_id TEXT NOT NULL, revision INTEGER NOT NULL DEFAULT 1, PRIMARY KEY(id))""")
            db.execSQL("CREATE UNIQUE INDEX IF NOT EXISTS summaries_date_idx ON daily_summaries(date)")
            db.execSQL("""CREATE TABLE IF NOT EXISTS app_metadata (id INTEGER NOT NULL DEFAULT 1, device_id TEXT NOT NULL, schema_version_value INTEGER NOT NULL, legacy_migration_version INTEGER NOT NULL DEFAULT 0, last_sync_at INTEGER, last_sync_status TEXT NOT NULL DEFAULT '', PRIMARY KEY(id))""")
        }

        private fun insertMetadata(
            db: SupportSQLiteDatabase,
            deviceId: String = UUID.randomUUID().toString(),
            legacyVersion: Int,
            schemaVersion: Int = 5
        ) {
            db.execSQL(
                "INSERT OR IGNORE INTO app_metadata (id, device_id, schema_version_value, legacy_migration_version, last_sync_status) VALUES (1, ?, ?, ?, '')",
                arrayOf<Any>(deviceId, schemaVersion, legacyVersion)
            )
        }
    }
}
